CI - no absolute links to local pages

aduermael opened the pull request:
### Describe the proposed changes

New tests to make sure there are no absolute links to `docs.docker.com` in local pages.
- <s>makes sure all local pages referenced by links do exist (and not though redirects)</s>
- <s>makes sure all redirects point to existing pages</s>

**edit:** I'll take care of relative links checking in a different PR.

**work in progress...**

johndmulhausen commented:
Deploy preview ready!

Built with commit 7c68919ac5bcbbe2a34aa23f1705347e58e31600

https://deploy-preview-849--docker-docs.netlify.com

aduermael commented:
Checks have passed! :)
@johndmulhausen can you merge it please?

I would like to start working on new tests for relative links to missing resources and potential redirect loops. 

Also, don't you think it's time to enforce tests on master? Almost all PRs to master branch with no Jenkinsfile have been merged. And I would gladly drop comments in the remaining ones to ask for rebase.

Thanks!

aduermael commented:
@johndmulhausen that PR is good to go! 😊

friism commented:
lgtm

aduermael commented:
@friism thanks! :)

johndmulhausen commented:
@aduermael Right now the CI is failing due to a submodules error:

```
20:57:32 Checking out Revision cde5c50f47419b74ff69e27a0281403082f6ef7a (PR-849)
20:57:33  > git config core.sparsecheckout # timeout=10
20:57:33  > git checkout -f cde5c50f47419b74ff69e27a0281403082f6ef7a
[Pipeline] sh
20:57:35 [ker_docker.github.io_PR-849-SIO4KRLIUPX4JYS5AYGAJID7R4JI6TFWDRPVZVH7ADS3DWRI6QYA] Running shell script
20:57:36 + git submodule update --init
20:57:36 No submodule mapping found in .gitmodules for path 'tests/src/github.com/gdevillele/frontparser'
[Pipeline] }
[Pipeline] // wrap
[Pipeline] }
[Pipeline] // wrap
[Pipeline] }
[Pipeline] // withDockerRegistry
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline

GitHub has been notified of this commit’s build result

ERROR: script returned exit code 1
Finished: FAILURE
```

aduermael commented:
@johndmulhausen Yes, Go deps are now vendored as submodules (you asked me to do it).
I don't know what's the right way to pull them using Jenkins. But I'll figure this out.

johndmulhausen commented:
It's great that we're gonna use submodules, thanks for that. Sorry it's creating this problem. Other than that, this PR is looking good. :) @mstanleyjones also did a pass at getting rid of old PRs so we should be able to turn on mandatory Jenkins very soon.

aduermael commented:
I fixed the issue with submodules, but meanwhile, new absolute links to docs.docker.com have been introduced... I'm fixing that.

aduermael commented:
Yay! All checks have passed! 🙂
@johndmulhausen, can you review it please?

aduermael commented:
@johndmulhausen ?

mstanleyjones commented on docker-for-windows/index.md:
```diff
@@ -82,7 +82,7 @@ enabled](troubleshoot.md#virtualization-must-be-enabled) in Troubleshooting.
 * Nested virtualization scenarios, such as running Docker for Windows on a VMWare or Parallels instance, might work, but come with no guarantees (i.e., not officially supported).
 <p />
-* **What the Docker for Windows install includes**: The installation provides [Docker Engine](https://docs.docker.com/engine/userguide/intro/), Docker CLI client, [Docker Compose](https://docs.docker.com/compose/overview/), and [Docker Machine](https://docs.docker.com/machine/overview/).
+* **What the Docker for Windows install includes**: The installation provides [Docker Engine](/engine/userguide/intro/), Docker CLI client, [Docker Compose](/compose/overview/), and [Docker Machine](/machine/overview/).
```
Not introduced here, but shouldn't this be a link to a .md file?

aduermael replied:
@mstanleyjones I have no idea... 😕

mstanleyjones commented on docker-for-windows/index.md:
```diff
@@ -478,15 +478,15 @@ You can configure options on the Docker daemon in the given JSON configuration f
 ![Docker Daemon](images/docker-daemon.png)
 
-For a full list of options on the Docker daemon, see <a href="https://docs.docker.com/engine/reference/commandline/dockerd/" target="_blank">daemon</a> in the Docker Engine command line reference.
+For a full list of options on the Docker daemon, see [daemon](/engine/reference/commandline/dockerd/) in the Docker Engine command line reference.
```
You changed the functionality of this link by removing the target attribute...

aduermael replied:
I can use this syntax with Kramdown: `[link](url){:target="_blank"}`
I'll fix it now.

mstanleyjones commented on docker-for-windows/index.md:
```diff
@@ -478,15 +478,15 @@ You can configure options on the Docker daemon in the given JSON configuration f
 In that topic, see also:
 
-* [Daemon configuration file](https://docs.docker.com/engine/reference/commandline/dockerd/#/daemon-configuration-file)
+* [Daemon configuration file](/engine/reference/commandline/dockerd/#/daemon-configuration-file)
```
Not introduced here, but these should point to the .md file if possible.

aduermael replied:
@mstanleyjones I can fix it if you're sure we should always link to md files

aduermael replied:
@mstanleyjones well if I do this the redirect loses the anchor link. You can try by yourself:

https://docs.docker.com/engine/reference/commandline/dockerd/#/daemon-configuration-file

versus

https://docs.docker.com/engine/reference/commandline/dockerd.md#/daemon-configuration-file

mstanleyjones replied:
The correct Markdown should be:

```markdown
[Daemon configuration file](/engine/reference/commandline/dockerd.md#daemon-configuration-file)
```

aduermael replied:
@mstanleyjones https://docs.docker.com/engine/reference/commandline/dockerd.md#daemon-configuration-file doesn't work either...

thaJeztah replied:
maybe it expects relative URLs; `../engine/reference/commandline/docker.md#daemon-configuration-file` ?

mstanleyjones commented on machine/concepts.md:
```diff
@@ -37,7 +37,7 @@ For a complete list of `docker-machine` subcommands, see the [Docker Machine sub
 Users using their own Docker Registry will experience `x509: certificate signed by unknown authority` 
 error messages if their registry is signed by custom root Certificate Authority and it is 
-not registered with Docker Engine. As discussed in the [Docker Engine documentation](https://docs.docker.com/engine/security/certificates/#/understanding-the-configuration)
+not registered with Docker Engine. As discussed in the [Docker Engine documentation](/engine/security/certificates/#/understanding-the-configuration)
```
As above, link to the .md if possible.

aduermael replied:
@mstanleyjones Same as above... Losing the anchor link

aduermael replied:
@mstanleyjones Also, one important point. Currently it is somehow better to use a path of this form: https://docs.docker.com/engine/reference/commandline/dockerd/
Instead of https://docs.docker.com/engine/reference/commandline/dockerd.md

Because the non-md form leads you right to the page you want. While the md form returns a terrible 404 error with a client side JS script that takes you to the non-md path...

We already mentioned that in the past, I really think it's an issue for SEO. 

But I also think it makes more sense to link to local markdowns within markdowns. Maybe we should just update these links at build time.

aduermael commented:
@mstanleyjones I added the missing `target=_blank`. And fixed some links to md files. 
But I can't fix them all, as the one containing anchor links would be broken. 

Please merge that PR as soon as possible, I don't want absolute links to be re-introduced.

aduermael commented:
@mstanleyjones @johndmulhausen Also, the Jenkins tests should be made required now. Please please please...

aduermael commented:
@mstanleyjones about that issue with links that should point to actual markdowns. I'm totally on board with this idea.
But I'm sure it happens in many files, and I would prefer to fix that in a dedicated PR, along with new tests for that particular issue. Because as of today, there's nothing making sure that will never happen again. 
So can you please merge that one @mstanleyjones? I'll submit a new one to address that issue. 
Thanks!
cc @thaJeztah @johndmulhausen

aduermael commented:
thank you @mstanleyjones 🙂
